File: JAVA/153 - Find Minimum in Rotated Sorted Array.java
/*
 153. Find Minimum in Rotated Sorted Array (M)
 
 Approach 1 : BS
 If the list isn't rotated, just return the first element.
 If rotated, compare the mid and low value and BS it based on the rotation.
 Time: O(logn)
 Space: O(1) 

*/

/* Approach 1 */
class Solution {
    public int findMin(int[] nums) {
        int low = 0, high = nums.length - 1;
        
        // If the list isn't rotated, return the first element
        if (nums[high] >= nums[low]) return nums[low];
        
        while (low < high - 1) {
            int mid = low + (high - low) / 2;
            // If mid > low value, then the smallest is in right
            if (nums[mid] > nums[low])
                low = mid;
            // else, in left
            else
                high = mid;
        }
        return Math.min(nums[low], nums[high]);
    }
}

File: C++/153-Find_Minimum_in_Rotated_Sorted_Array.cpp
/*
 153. Find Minimum in Rotated Sorted Array
 Medium
*/

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        return findMin(nums, l, r);
    }
 
    int findMin(vector<int>& nums, int l, int r)
    {
        if (l + 1 >= r)
            return min(nums[l], nums[r]);
 
        if (nums[l] < nums[r])
            return nums[l];
 
        int m = l + (r - l) / 2;
 
        return min(findMin(nums, l, m - 1), findMin(nums, m, r));
    }
};
